docs/design/robots.md

```diff
@@ -0,0 +1,198 @@
+#Service Accounts
+
+## Motivation 
+
+Processes in Pods may need to call the Kubernetes API.  For example:
+  - scheduler
+  - replication controller
+  - minion controller
+  - a map-reduce type framework which has a controller that then tries to make a dynamically determined number of workers and watch them
+  - continuous build and push system
+  - monitoring system
+
+Distinguishing between humans and service accounts (a.k.a. robot accounts or roles) is a practice used by a number of cloud providers, such as:
+    - Google [service accounts](https://developers.google.com/accounts/docs/OAuth2ServiceAccount), which can be attached
+      to VM instances.
+    - Amazon Web Services [Roles](http://docs.aws.amazon.com/IAM/latest/UserGuide/Using_CreateRole_CLIAPI.html), which
+      can be attached to VM instances.
+    - quay.io [robot accounts](http://docs.quay.io/glossary/robot-accounts.html)
+
+The requirements for humans and automated processes are different:
+    - Humans need a wide range of capabilities to do their daily activities. Automated processes often have more narrowly-defined activities.
+    - Humans may better tolerate the exceptional conditions created by expiration of a token.  Remembering to handle
+      this in a program is more annoying.  So, either long-lasting credentials or automated rotation of credentials is
+      needed.
+    - A Human typically keeps credentials on a machine that is not part of the cluster and so not subject to automatic
+      management.  A VM with a role/service-account can have its credentials automatically managed.
+
+Therefore, Kubernetes should support both human and service accounts, with different ways of managing credentials and
+policies for each.
+
+## Design Overview
+
+A new REST object is added to the apiserver that represents a service account, called `serviceAccount`.  Another new
+REST object is added to the apiserver that represents a secret which an account needs to prove its identity to peers.
+This is called `secret`.  A `serviceAccount` has access to one or more `secret` objects.  The storage of secrets is separate to allow listing `/serviceAccounts` without seeing secrets.
+
+The `PodSpec` is extended to allow Pods to be started with an associated serviceAccount.  
```

Comment:
Though experiment:

* sooner or later, every process in a cluster has to run in some kind of security context
* the reality of Unix is that uid/gid and to a lesser extent selinux mcs labels are the only mechanisms in a cluster for identifying the difference between security contexts in the kernel, process space, and filesystem
* it follows that every process run by Kubernetes at a node level should be assigned to at least one security context
* security contexts should ideally be isolated across namespaces, but no process should be run outside of a security context
* an auditor on a host has no option in some cases but to attempt to associate a process uid/gid to higher level account, and real world large scale clusters reflect this by avoiding reusing uids/gids across unrelated services
* to me, the secrets available to a process constitute part of a security context, since audit must be able to reason about the capabilities a process has on other systems

If the system has a default security context, all pods would run under that.  If a namespace has a default security context, it's pods would run under that.  There may be a desire to run individual pods in a namespace within different contexts.  However, both of the first two options do not require pods to be aware of a security context, just the kubelet.  A service account with secrets might be one mechanism for that security context to be configured, but does not imply that is the only mechanism.